Load REDIS_VERSION from ENV_DIR and not the environment

edmorley opened the pull request:
Since the buildpack API doesn't populate the environment automatically, we need to do so ourselves (follow up to #11):
https://devcenter.heroku.com/articles/buildpack-api#bin-compile

This follows the approach used by `heroku-buildpack-ci-postgresql`:
https://github.com/heroku/heroku-buildpack-ci-postgresql/blob/54c6d39bf997ebb612f7db07aab98cde1c5777c6/bin/compile#L22-L26

Also performs some compile output and CI-related cleanup.

See commits for details.

Fixes [W-6167563](https://gus.my.salesforce.com/a07B0000006toMRIAY).

edmorley committed:
Fix handling of REDIS_VERSION

Since the buildpack API doesn't populate the environment automatically, we need to do so ourselves (follow up to #11): https://devcenter.heroku.com/articles/buildpack-api#bin-compile.

This follows the approach used by `heroku-buildpack-ci-postgresql`: https://github.com/heroku/heroku-buildpack-ci-postgresql/blob/54c6d39bf997ebb612f7db07aab98cde1c5777c6/bin/compile#L22-L26.

The tests didn't originally catch this since the `ARG REDIS_VERSION` was being propagated into the compile script. Now we are cleaning the environment variables they do.

edmorley committed:
Add a short sleep before running `redis-cli info` in tests

To reduce the chance of intermittent test failures due to redis-server not having quite finished starting up. eg: https://travis-ci.com/heroku/heroku-buildpack-ci-redis/jobs/201965047#L1410.

bensymonds committed:
Synchronise compile output with equivalent postgres CI buildpack

Mainly this ensures the used redis version is displayed. But also it's nice if the other output is consistent too.

bensymonds committed:
Capitalise buildpack name

This makes it more consistent with other buildpacks.

bensymonds commented on bin/test.sh:
```diff
@@ -31,7 +31,7 @@ echo "Checking redis-server responds to ping command..."
 # https://stackoverflow.com/questions/38271281/can-i-use-redis-cli-with-a-connection-url
 REDIS_CONNECTION_ARGS="\$(echo \${REDIS_URL} | sed 's_redis://h:\(.*\)@\(.*\):\(.*\)/_-h \2 -p \3 -a \1_')"
-TEST_COMMAND="source .profile.d/redis.sh && redis-cli ${REDIS_CONNECTION_ARGS} ping"
+TEST_COMMAND="source .profile.d/redis.sh && sleep 1 && redis-cli ${REDIS_CONNECTION_ARGS} ping"
```
`sleep`s always feel a bit hack to my. I wonder if the `-r -i` options to `redis-cli` might be better?

jabrown85 replied:
I like that idea too @bensymonds. Something like `-r 5 -i 200`

bensymonds replied:
@jabrown85 The problem I see with those values is if something has gone catastrophically wrong and redis is not going to be ready _at all_, the build will take ~1000 seconds (5 * 200) to give up trying, which is a long time to wait for a CI build to finish. Given @edmorley had gone with `sleep 1` I'm assuming it's only a really short delay before it's ready, so I went with a more optimistic `-r 5 -i 2`. What do you think?

jabrown85 replied:
Sorry, yeah I thought it was in milliseconds.

bensymonds replied:
Ah ok, haha. I thought you were just an extremely patient person.

bensymonds commented on Dockerfile:
```diff
@@ -4,10 +4,12 @@ ARG RUNTIME_IMAGE
-ARG REDIS_VERSION
+# Using a prefixed variable name, to prevent the compile from accidentally relying on it
+# instead of remembering to load in /env/REDIS_VERSION as required by the buildpack API.
+ARG TEST_REDIS_VERSION
```
I think we also need to update `test.sh` to pass this new name. At the moment we haven't. The result is that the `TEST_REDIS_VERSION` from the env matrix is being ignored, so we are installing the default redis version in every build, and so the build is still green. I wonder if we can make the build fail if a typo like this happens again?

bensymonds replied:
(For anyone reading this - these are notes to myself given @edmorley is away, so not expecting a response. But feel free to comment if you have something to say!)

bensymonds replied:
We now check the redis version that ended up being installed, which covers this.

jabrown85 commented on .travis.yml:
```diff
@@ -2,14 +2,16 @@ language: bash
+  - STACK=heroku-16 TEST_REDIS_VERSION=3
+  - STACK=heroku-16 TEST_REDIS_VERSION=4
+  - STACK=heroku-16 TEST_REDIS_VERSION=5
+  - STACK=heroku-16 TEST_REDIS_VERSION=3
```
stack should be `heroku-18` here

bensymonds replied:
Ah good spot :+1:

bensymonds commented on bin/test.sh:
```diff
@@ -31,7 +31,7 @@ echo "Checking redis-server responds to ping command..."
 # https://stackoverflow.com/questions/38271281/can-i-use-redis-cli-with-a-connection-url
 REDIS_CONNECTION_ARGS="\$(echo \${REDIS_URL} | sed 's_redis://h:\(.*\)@\(.*\):\(.*\)/_-h \2 -p \3 -a \1_')"
-TEST_COMMAND="source .profile.d/redis.sh && redis-cli ${REDIS_CONNECTION_ARGS} ping"
+TEST_COMMAND="source .profile.d/redis.sh && redis-cli -r 5 -i 2 ${REDIS_CONNECTION_ARGS} ping"
```
This doesn't work, it still fails on the first attempt. Makes sense when hindsight - this is for _repeating_ a command, not _retrying_. Instead I'll look at using something (`nc`?) to wait for the port to open. And if _that_ doesn't work then go back to the `sleep` 😄

bensymonds replied:
Couldn't see a clean way of doing it so sticking with the `sleep` for now.

bensymonds committed:
Check expected version of redis was installed in tests

To cover us against situations where the default version is accidentally being built instead of a specified version.

bensymonds committed:
Remove usage of runtime image in tests

Heroku CI dynos use the "build" image variant.

bensymonds committed:
Update dockerfile to more closely emulate real environment

As per the Buildpack API (https://devcenter.heroku.com/articles/buildpack-api)
the buildpack scripts are run in a environment stripped of most
environment variables - emulate that here by using `env -i`.

In real life there is a whitelist of env vars. Rather than put them all
in, by trial-and-error I determined only PATH and HOME are required.
It feels a bit hacky to do it like that but I think it's ok for now.

bensymonds commented:
@edmorley I messed around with this quite a bit in your absence 😀(sorry). Take a look and let me know what you think.

edmorley committed:
Test all permutations of Stack and Redis versions

Whilst the test matrix here is large and there is inevitably a lot of
overlap, this repository has low traffic and a fast CI run, so there's
little to be gained by skipping some of the version permutations.

bensymonds reviewed (approved):
Looks 💯
